geth/peers/peerpool.go

```diff
@@ -125,47 +165,51 @@ func (p *PeerPool) restartDiscovery(server *p2p.Server) error {
 
 // handleServerPeers watches server peer events, notifies topic pools about changes
 // in the peer set and stops the discv5 if all topic pools collected enough peers.
+//
+// @TODO(adam): split it into peers and discovery management loops. This should
+// simplify the whole logic and allow to remove `timeout` field from `PeerPool`.
 func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.PeerEvent) {
 	var retryDiscv5 <-chan time.Time
 
 	for {
+		p.mu.RLock()
+		timeout := p.timeout
```

Comment:
why you can't set timeout once? who can change it?

Reply:
Because Discovery may be stopped due to reaching max peers limit. Currently, there are two scenarios: (1) `PeerPool` stops by calling `Stop()` and Discovery together with all loops are closed or (2) it stops due to reaching max peers limit and in this case only Discovery is stopped, not the `handleServerPeers` loop. This is because if a peer is dropped, discovery is restarted and a new timeout must be created.